### PR TITLE
Allow Configuration to handle nested structures and lists

### PR DESCRIPTION
**Pull request by davidgraeff**
One step towards #4178

The `Configuration` class provides unmarshalling to a POJO class. The `as` method is extended to support nested structures and list of structures.

An example configuration class structure that benefits from nested unmarshalling and from list unmarshalling:
```java
class ServiceConfiguration {
   BigDecimal importantNumber;
   String niceString;
   CoordinateConfiguration theCoordinates;
   List<UserConfiguration> users;
}

class CoordinateConfiguration {
   BigDecimal x;
   BigDecimal y;
}
class UserConfiguration {
   String username;
}
```

## Port of tests to java
I ported some relevant tests to java and added a test case for nested and list fields

## Marshal class back to Map<String, Object>
Add the method `Map<String,Object> readFromConfigurationClassInstance(Object configurationClassInstance) `

## Discussion
* I marked `ConfigUtil:normalizeType(Object value)` deprecated in favour of `ConfigUtil:normalizeType(Object value, @Nullable ConfigDescriptionParameter)`. In my opinion with the deprecated method we are hiding the fact, that we always create a Normalizer in that specific method in a `null` case. The consumer should specifically confirm with null that he will not provide a config description and happily accepts a created Normalizer. An alternative is to create a default Normalizer in the constructor. In that case `ConfigUtil:normalizeType(Object value)` is a valid method signature again.
* I removed the exception that is thrown in `ConfigUtil:normalizeType` (https://github.com/eclipse/smarthome/pull/4181/files#diff-810126e50ad4f2c2ad5d985ffd577ed4L87) if the object is not a String or a Number. It was a non checked exception, not documented and not in the unit tests and not part of the API, so I guess nothing depends on this exception to be thrown. Instead the value object is just returned as-is. This change is necessary to allow Map and List objects.

### CODE REVIEW

**Comment by sjka** [comment]
Just a (very) minor remark: `Configuration` is still using the newly deprecated method signature:

https://github.com/davidgraeff/smarthome/blob/4faa4b1b5d1a77d72a080292f04dc8f284dd469b/bundles/config/org.eclipse.smarthome.config.core/src/main/java/org/eclipse/smarthome/config/core/Configuration.java#L234

Otherwise this looks great!

> I marked ConfigUtil:normalizeType(Object value) deprecated in favour of [...]

👍 

> I removed the exception that is thrown in ConfigUtil:normalizeType  [...] if the object is not a String or a Number. [...] Instead the value object is just returned as-is. 

I'm also not aware of any code "relying" on it. The only aspect I can think of is that it pointed out beautifully that there something was seriously wrong. On the other hand: there is still something going to explode nicely downstream if the types don't match, so I'd be okay to leave it as it is now and see how it goes.

**Comment by sjka** [comment]
Could you please rebase this PR?

File: bundles/config/org.eclipse.smarthome.config.core/src/main/java/org/eclipse/smarthome/config/core/Configuration.java [hunk 1]
```diff
@@ -53,63 +56,143 @@ public Configuration(Map<String, Object> properties) {
...
+    /**
+     * Creates an instance of the given configuration class (a default constructor is required) and
+     * tries to find a matching property value for each non-final field.
+     * Nested classes are supported.
+     * List fields are supported as long as the property value is of type Collection.
+     *
+     * @param configurationClass
+     * @return
+     */
+    @Nullable
```
Comment (maggu2810) [inline bundles/config/org.eclipse.smarthome.config.core/src/main/java/org/eclipse/smarthome/config/core/Configuration.java]: Type annotations should be placed at the type and not on the visible modifier

File: bundles/config/org.eclipse.smarthome.config.core/src/main/java/org/eclipse/smarthome/config/core/Configuration.java [hunk 2]
```diff
@@ -53,63 +56,143 @@ public Configuration(Map<String, Object> properties) {
...
+    }
+
+    /**
+     * Creates an instance of the given configuration class (a default constructor is required) and
+     * tries to find a matching property value for each non-final field.
+     * Nested classes are supported.
+     * List fields are supported as long as the property value is of type Collection.
+     *
+     * @param configurationClass
+     * @return
```
Comment (maggu2810) [inline bundles/config/org.eclipse.smarthome.config.core/src/main/java/org/eclipse/smarthome/config/core/Configuration.java]: an instance of the given class on success, null on error

**Comment by kaikreuzer** [comment]
> Otherwise this looks great!

@sjka This pretty much contradicts [your comment here](https://github.com/eclipse/smarthome/issues/3692#issuecomment-317672269) - although I was specifically talking about Maps, the general theme was the same: Having more complex structures as configurations.

The big issue with the suggestions (mine as well as David's) is that we do not stand a chance to provide proper config descriptions for them, nor to have it in any way supported in the UIs - and that was the ultimate goal of all the configuration infrastructure.

So although I'd personally like to be able to have more complex configurations, I do not see a way to move away from flat lists of primitive types. Do I miss anything?

**Comment by kaikreuzer** [comment]
Hm, ok, I just read over https://github.com/eclipse/smarthome/issues/4178 and I see that I completely lost track of all activities over the last week - my suggestion would be to first discuss the general vision on #4178 and only process the individual PRs afterwards.

Ok for everyone?

**Comment by davidgraeff** [comment]
That was the plan. The PRs are proof of concept. It is possible to allow complex structures in the core. I haven't check the GUI situation yet. 

Another point to analyse is, how this would influence thing configuration. Because the dtd is shared between every way of configuration, things would be capable of having complex configs as well. This would need to be supported by storage, or the config needs to be flattened at some point.

**Comment by davidgraeff** [comment]
I will alter this PR, and add annotations for configuration fields. Annotations that I have in mind are:
* FieldRequired
* SecretVaultField
* Base64Chars
* Base64Bytes

**Comment by htreu** [comment]
@davidgraeff is there any activity planed towards your latest proposal for this PR? Please give a short update.

**Comment by davidgraeff** [comment]
Can be closed.

**Pull request closed without merging**